### DISCUSSION TITLE
Bash Scripting in Automation

### DISCUSSION THREAD

**Discussion by gpjorgenson** [Budibase/budibase#7447] [Help] [answered]
Like some others have mentioned, I cannot find any documentation or examples on bash scripting. 

I need to run a perl script when a button requests an automation...

but even something as simple as just a " ls /usr/work-gpj/custserv/ " has an output of:   "stdout": "Command failed: ls /usr/work-gpj/custserv/\n\nls: cannot access '/usr/work-gpj/custserv/': No such file or directory\n",
  "success": false
I have tried it with and without the #!/bin/bash 

Can anyone tell me how to get a bash scripting in Automation to work?

**Comment by R2bEEaton** [comment, marked as answer]
Hi @gpjorgenson!

The reason your `ls` command didn't work as expected is because the bash automation executes inside the Budibase container, not on the host. That's why if you do `pwd` (print working directory), you get `/app`.

I found [this answer](https://stackoverflow.com/a/20816397) on Stack Overflow for how to explore Docker container file systems (assuming you're using the Budibase Docker install) and was able to run a perl script from within Budibase.

### How to do it

1. Run `docker ps` to list your running containers.
2. If using the multiple container Budibase install, look for `bbapps` and copy the `CONTAINER ID`.
3. Run `docker exec -t -i CONTAINER_ID /bin/bash`, of course inserting your container ID. You have now entered the shell of your container.
4. Run `pwd` to ensure you are in the `/app` directory, same as when you run `pwd` in Budibase.
5. Create your perl program in this directory. I used this command to make a Hello World program: `echo 'print("Hello world!");' > test.pl`.
6. In your Budibase automation, create a Bash Scripting action and set the code to `perl test.pl`.

![image](https://user-images.githubusercontent.com/34921506/186644350-eb817a2b-adb0-4ac7-93d8-773cd3267a5e.png)
As you can see from the Test Details, this worked perfectly!

### Note

1. You could also just run the perl code in the Bash automation action using `perl -e 'print("Hello world!");'
`. If you can get your whole program in there, that's perfect and much easier than the above procedure.
2. Please note that this bash shell is incredibly limited. It doesn't even come with `curl` or `wget`, so depending on what you want to do, it may be impossible.
![image](https://user-images.githubusercontent.com/34921506/186644589-1eaf7f38-4312-4be3-a83b-934ce6bbfc10.png)
2. I also don't know what happens to files in this directory when you update Budibase. Make sure to backup your perl script just in case it gets deleted.

### Enjoy!
Reply with any questions or accept the answer using the check mark above to help future readers.

**Reply by gpjorgenson** [reply]
Thank you!   I will try this asap and report on the results. I need to check with our guy who maintains docker before I do.

**Reply by melohagan** [reply]
This is an amazing answer @R2bEEaton thank you :)

To be honest I've wondered how the bash scripting worked, as I've never used it myself. I can add a little documentation if I can figure it out for myself, the only thing is that I'm not really sure what the use case is.

@gpjorgenson What is you use case?

**Reply by gpjorgenson** [reply]
My use case was to initiate a perl script via bash scripting. The script will look at one postgresql database and copy selected fields into a second postgresql database. We have a stand alone instance of budibase using docker. We set up a docker bind volume linked to a local directory. A simple perl script ran fine but the more complicated perl script with multiple modules would return the  
{
 "stdout": "spawnSync /bin/sh ETIMEDOUT",
 "success": false
}
error. 
We extended the bbapps image by installing the additional perl distribution packages.
Still received the error when testing. 
We adjusted the QUERY_THREAD_TIMEOUT  to 30000ms but the error still remained.
I did try it after publishing and it indicated the automation ran but no tables were updated.
So I tried running the script inside the docker container and it ran just fine. Updated the SQL tables.
To troubleshoot, . I took the simple program that runs successfully thru budibase and started adding the use <module> lines one at a time.
use 5.10.1;
use open ":std", ":encoding(UTF-8)";
use DBI;
use Time::HiRes qw(time);
use POSIX qw(strftime);   all added in okay
but then I added
use DateTime qw( ); and got the error.
I was going to try adding the modules in different orders, numbers etc and see what else happens. 
But, at this point, it was decided that our docker tech proposed putting mr perl work behind a REST api that budibase can call, instead of invoking perl directly.  That way there is no "timeout" to worry about.
That would also decouple perl from budibase, so the bbapp container can just be stock, unmodified.
We are going to try what he proposed.

**Reply by gpjorgenson** [reply]
Oh, for the use case, I am not allowed to have the first database inside the budibase due to some HIPAA security requirements. If I could have had all the tables inside the budibase app, it would have been a simple to accomplish what was needed but, alas, that was not allowed.

**Reply by R2bEEaton** [reply]
> Oh, for the use case, I am not allowed to have the first database inside the budibase due to some HIPAA security requirements. If I could have had all the tables inside the budibase app, it would have been a simple to accomplish what was needed but, alas, that was not allowed.

Interesting. Glad you guys are on your way to a solution that works for you!

**Reply by melohagan** [reply]
Hmm, but if it's an external PostgreSQL connection, the data wouldn't be stored in Budibase, it would be stored in the Postgres DB...

If the need is to only show specific fields, would views have worked? (If you could use them that is: https://github.com/Budibase/budibase/issues/6216)

**Reply by gpjorgenson** [reply]
Customer service requires tables they can alter without affecting the original data. Hence, the 2nd db. Plus, it has been decided the 2nd DB will be on a different server than the first db. Special security requirements exist.
